daemon: change default policy enforcement behavior for non-K8s envs

ianvernon opened the pull request:
Previously, in an environment where Cilium was not running in tandem with
Kubernetes, policy enforcement would be enabled for all endpoints if a  rule
was added to the policy repository even if said rule did not apply to an
endpoint. This changes that behavior for Cilium running without Kubernetes to
enable or disable policy enforcment as is done when Cilium is ran with
Kubernetes; policy enforcement is only enabled for an endpoint when a rule
selects that endpoint when the daemon is ran in 'default' policy enforcement
mode.

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes #1424

tgraf reviewed (changes requested):
Changes look great, please add a section to the docs to describe this behaviour:
http://cilium.readthedocs.io/en/latest/policy/#

Commit:
tests: update 01-ct.sh to reflect policy changes

Update the test so that it takes into account the changes in policy enforcement
in non-K8s setups.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
daemon: change default policy enforcement behavior for non-K8s envs

Previously, in an environment where Cilium was not running in tandem with
Kubernetes, policy enforcement would be enabled for all endpoints if a  rule
was added to the policy repository even if said rule did not apply to an
endpoint. This changes that behavior for Cilium running without Kubernetes to
enable or disable policy enforcment as is done when Cilium is ran with
Kubernetes; policy enforcement is only enabled for an endpoint when a rule
selects that endpoint when the daemon is ran in 'default' policy enforcement
mode.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
tests: remove 15-policy-config.sh

* The overlap of 15-policy-config.sh with 14-policy-enforcement-docker.sh is
considerable, and 14-policy-enforcement-docker.sh does much more extensive
testing than 15-policy-config did for validating the state machine for policy
enforcement.
* Moved the testing of whether policy trace functionality works
correctly in the case where no policy is imported to 12-policy-import.sh.
* Moved the "ping_success" and "ping_fail" functions from 15-policy-config.sh
to helpers.bash as they can be leveraged elsewhere in the code.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
tests: update wait_for_policy_enforcement

Make wait_for_policy_enforcement take an argument for how many endpoints should
have policy enforcement disabled.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
Documentation: update policy documentation

Update policy documentation to include information about Policy Enforcement modes.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
tests: remove unneeded helper call from 10-proxy

We do not need to invoke the wait_for_policy_enforcement function in this test;
policy enforcement is enabled when an endpoint is regenerated / and it goes to
ready state which is already checked prior in the test.

Signed-off by: Ian Vernon <ian@cilium.io>
